### PR TITLE
[FIX] account: Payment matching with cash basis

### PR DESCRIPTION
**Pull request by simongoffin**
Steps to reproduce the bug:

- Let's consider the current company CY in €
- Let's consider a cash basis tax T of 20%
- Let's consider that 1€ = 2$
- Create a customer invoice I on journal J in € for customer C
- Create a line on J with 100€ and set T on it
- Post I
- Create a payment P of 240$ for C on bank journal in €
- Confirm P and click on payment matching
- Try to reconcile P with I

Bug:

A traceback was raised.

opw:2256207

### CODE REVIEW

File: addons/account/models/account_move.py [hunk 1]
```diff
@@ -4321,8 +4321,8 @@ def create_tax_cash_basis_entry(self, percentage_before_rec):
                                 'tax_exigible': True,
                                 'tax_ids': [(6, 0, [tax.id])],
                                 'move_id': newly_created_move.id,
-                                'currency_id': line.currency_id.id,
-                                'amount_currency': self.amount_currency and line.currency_id.round(line.amount_currency * amount / line.balance) or 0.0,
+                                'currency_id': self.currency_id.id,
+                                'amount_currency': self.amount_currency and self.currency_id.round(-self.amount_currency * amount / line.balance) or 0.0,
```
Comment (smetl) [inline addons/account/models/account_move.py:4325]: replacing `line` by `self` seems completely wrong...

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -4336,8 +4336,8 @@ def create_tax_cash_basis_entry(self, percentage_before_rec):
                                 'account_id': account_id.id,
                                 'tax_exigible': True,
                                 'move_id': newly_created_move.id,
-                                'currency_id': line.currency_id.id,
-                                'amount_currency': self.amount_currency and line.currency_id.round(-line.amount_currency * amount / line.balance) or 0.0,
+                                'currency_id': self.currency_id.id,
+                                'amount_currency': self.amount_currency and self.currency_id.round(self.amount_currency * amount / line.balance) or 0.0,
```
Comment (smetl) [inline addons/account/models/account_move.py:4340]: the sign `-` disappeared, is it intented?

File: addons/account/models/account_move.py (hunk 1 above)
Comment (simongoffin) [reply addons/account/models/account_move.py:4325]: amount_currency is equal to self.amount_currency and not to line.amount_currency

File: addons/account/models/account_move.py (hunk 2 above)
Comment (simongoffin) [reply addons/account/models/account_move.py:4340]: Yes it's intended as this created aml is a credit, the sign - must be removed as it is already included in the balance (debit - credit)

File: addons/account/models/account_move.py (hunk 1 above)
Comment (smetl) [reply addons/account/models/account_move.py:4325]: `'amount_currency': line.currency_id.round(line.amount_currency * amount / line.balance) if line.currency_id and line.balance else 0.0,`

File: addons/account/models/account_move.py (hunk 2 above)
Comment (smetl) [reply addons/account/models/account_move.py:4340]: `'amount_currency': line.currency_id.round(-line.amount_currency * amount / line.balance) if line.currency_id and line.balance else 0.0,
`

**Comment by simongoffin** [comment]
robodoo r+

File: addons/account/models/account_move.py [hunk 3]
```diff
@@ -4322,8 +4322,8 @@ def create_tax_cash_basis_entry(self, percentage_before_rec):
                                 'tax_exigible': True,
                                 'tax_ids': [(6, 0, [tax.id])],
                                 'move_id': newly_created_move.id,
-                                'currency_id': line.currency_id.id,
-                                'amount_currency': self.amount_currency and line.currency_id.round(line.amount_currency * amount / line.balance) or 0.0,
+                                'currency_id': self.currency_id.id,
```
Comment (smetl) [inline addons/account/models/account_move.py:4325]: this self is wrong...

File: addons/account/models/account_move.py [hunk 4]
```diff
@@ -4337,8 +4337,8 @@ def create_tax_cash_basis_entry(self, percentage_before_rec):
                                 'account_id': account_id.id,
                                 'tax_exigible': True,
                                 'move_id': newly_created_move.id,
-                                'currency_id': line.currency_id.id,
-                                'amount_currency': self.amount_currency and line.currency_id.round(-line.amount_currency * amount / line.balance) or 0.0,
+                                'currency_id': self.currency_id.id,
```
Comment (smetl) [inline addons/account/models/account_move.py:4340]: this self is wrong...

**Pull request closed without merging**